### PR TITLE
Feature/rework api code generation

### PR DESCRIPTION
**Pull request by piiertho**
This enables api code generation using kotlin poet.
The resulting code is more cleaner, like only needed classes and members are imported in generated classes.
It has been tested on the sample projects running on Godot 3.0.6 as we have not yet update api to 3.1
This solves #14

### CODE REVIEW

**Comment by piiertho** [comment]
Note : When we'll update api to 3.1 we should use `godot.json` that has been added to api repo.

File: tools/api-classes-generator/src/main/kotlin/Property.kt [hunk 1]
```diff
@@ -34,8 +35,59 @@ class Property(
...
+                            .addStatement("%M(${validSetter.name}MethodBind, this.rawMemory${if (index != -1) ", $index, value)" else ", value)"}",
+                                    MemberName("godot.icalls", icall.name))
+                            .build()
+            )
+        }
+        if (hasValidGetter) {
+            val icall = if (index != -1)
+                ICall(type, listOf(Argument("idx", "Long")))
+            else
+                ICall(type, listOf())
```
Comment (chippmann) [inline tools/api-classes-generator/src/main/kotlin/Property.kt]: curly braces

File: tools/api-classes-generator/src/main/kotlin/Property.kt [hunk 2]
```diff
@@ -34,8 +35,59 @@ class Property(
...
+            icalls.add(icall)
+            propertySpecBuilder.getter(
+                    FunSpec.getterBuilder()
+                            //Hard to maintain but do not see how to do better (Pierre-Thomas Meisels)
+                            .addStatement("return %M(${validGetter.name}MethodBind, this.rawMemory${if (index != -1) ", $index)" else ")"}",
+                                    MemberName("godot.icalls", icall.name))
```
Comment (chippmann) [inline tools/api-classes-generator/src/main/kotlin/Property.kt]: each param on new line and closing bracket on new line

File: tools/api-classes-generator/src/main/kotlin/Property.kt [hunk 3]
```diff
@@ -34,8 +35,59 @@ class Property(
...
+                            .build()
+            )
+        }
+        else propertySpecBuilder.getter(
```
Comment (chippmann) [inline tools/api-classes-generator/src/main/kotlin/Property.kt]: curly braces

File: tools/api-classes-generator/src/main/kotlin/Property.kt [hunk 4]
```diff
@@ -34,8 +35,59 @@ class Property(
...
+                FunSpec.getterBuilder()
+                        .addStatement("%L %T(%S)", "throw", UninitializedPropertyAccessException::class, "Cannot access property $name: has no getter")
```
Comment (chippmann) [inline tools/api-classes-generator/src/main/kotlin/Property.kt]: each param on new line and closing bracket on new line

File: tools/api-classes-generator/src/main/kotlin/Property.kt [hunk 5]
```diff
@@ -34,8 +35,59 @@ class Property(
...
+                        .build()
+        )
+
+        return propertySpecBuilder.build()
+    }
+
+    infix fun applyGetterOrSetter(method: Method) {
```
Comment (chippmann) [inline tools/api-classes-generator/src/main/kotlin/Property.kt]: i would also cleanup code in this function regarding formatting

File: tools/api-classes-generator/src/main/kotlin/TypeCast.kt [hunk 6]
```diff
@@ -26,66 +58,41 @@ fun String.removeEnumPrefix(): String {
     return ret.replace("::", ".").escapeUnderscore()
 }
 
-
-
+fun String.getPackage() =
+        when {
+            isEnum() -> {
+                var ret = this
+                val ind = ret.indexOf("enum.")
```
Comment (chippmann) [inline tools/api-classes-generator/src/main/kotlin/TypeCast.kt]: what means ret and ind?
personally i don't like shortened variable, parameter or class names. this is exactly a case where i ask myself what the shortened version means ^^

File: tools/api-classes-generator/src/main/kotlin/TypeCast.kt [hunk 7]
```diff
@@ -26,66 +58,41 @@ fun String.removeEnumPrefix(): String {
...
+                if (ind != -1)
+                    ret = ret.drop(ind + 5)
+
+                if (ret == "Error")
```
Comment (chippmann) [inline tools/api-classes-generator/src/main/kotlin/TypeCast.kt]: curly braces

File: tools/api-classes-generator/src/main/kotlin/Argument.kt [hunk 8]
```diff
@@ -22,6 +24,18 @@ class Argument(
...
+                "Color", "Variant" -> "$type($defaultValue)"
+                "Boolean" -> defaultValue.toLowerCase()
+                "Double" -> intToFloat(defaultValue)
+                "Vector2", "Vector3", "Rect2" -> "$type$defaultValue"
+                "Transform", "Transform2D", "GDArray", "RID", "PoolVector2Array", "PoolStringArray", "PoolVector3Array", "PoolColorArray", "PoolIntArray", "PoolRealArray", "PoolByteArray" -> "$type()"
+                "String" -> "\"$defaultValue\""
+                else -> defaultValue
+            }
+        } else null
     }
```
[the remark this answers is not in the document]
Comment (piiertho) [reply tools/api-classes-generator/src/main/kotlin/Argument.kt]: indeed, forgot to remove it.

File: tools/api-classes-generator/src/main/kotlin/Class.kt [hunk 9]
```diff
@@ -34,129 +39,178 @@ class Class(
...
-
-
-                if (isSingleton)
-                    append("    @ThreadLocal") // TODO: remove later, fixed in konan master
-                appendln("    companion object {")
-
-                append(generateCasts(tree))
-
-                constantsPrefix = "        "
+        properties.forEach { propertie ->
```
[the remark this answers is not in the document]
Comment (piiertho) [reply tools/api-classes-generator/src/main/kotlin/Class.kt]: Pardon my french xD

**Pull request merged**